pkg/controller/workloadspread/workloadspread_controller.go

```diff
@@ -391,35 +397,66 @@ func (r *ReconcileWorkloadSpread) getSuitableSubsetForOldPod(ws *appsv1alpha1.Wo
 			klog.Errorf("unexpected error occurred when matching pod (%s/%s) with subset, please check requiredSelectorTerm field of subset (%s) in WorkloadSpread (%s/%s), err: %s",
 				pod.Namespace, pod.Name, subset.Name, ws.Namespace, ws.Name, err.Error())
-		} else if matched {
-			value, _ := json.Marshal(&wsutil.InjectWorkloadSpread{
-				Name:   ws.Name,
-				Subset: subset.Name,
-			})
-			patch, _ := json.Marshal(map[string]interface{}{
-				"metadata": map[string]interface{}{
-					"annotations": map[string]string{
-						wsutil.MatchedWorkloadSpreadSubsetAnnotations: string(value),
-					},
-				},
-			})
-			if err = r.Patch(context.TODO(), pod, client.RawPatch(types.StrategicMergePatchType, patch)); err != nil {
-				// no need to retry, this operation will be retried in the next reconcile
-				// this failure would not affect the subsequent logic
-				klog.Errorf(`Failed to patch "matched-workloadspread:{Name: %s, Subset: %s} annotation" for pod (%s/%s), but will retry in the next reconcile, err: %s`,
-					ws.Name, subset.Name, pod.Namespace, pod.Name, err.Error())
+		}
+		// select the most favorite subsets for the pod by matching subset.PreferredNodeSelectorTerms
+		if matched && preferredScore > maxPreferredScore {
+			favoriteSubset = subset
+			maxPreferredScore = preferredScore
+		}
+	}
+
+	// patch MatchedWorkloadSpreadSubsetAnnotations to the pod;
+	// select labels/annotations form favoriteSubset.patch, and patch them to the pod;
+	if favoriteSubset != nil {
+		var patchLabels map[string]string
+		patchAnnotations := make(map[string]string)
+		patchMetadata := make(map[string]interface{})
+
+		// decode favoriteSubset.patch.raw and add their labels/annotations to the patch
+		if favoriteSubset.Patch.Raw != nil {
+			patchField := &corev1.Pod{}
+			if err := json.Unmarshal(favoriteSubset.Patch.Raw, patchField); err == nil {
+				patchLabels = patchField.Labels
+				if patchField.Annotations != nil {
+					patchAnnotations = patchField.Annotations
+				}
 			}
-			return subset, nil
 		}
+
+		// add MatchedWorkloadSpreadSubsetAnnotations to the patch
+		value, _ := json.Marshal(&wsutil.InjectWorkloadSpread{
+			Name:   ws.Name,
+			Subset: favoriteSubset.Name,
+		})
+		patchAnnotations[wsutil.MatchedWorkloadSpreadSubsetAnnotations] = string(value)
+
+		// construct the patch expression, and patch it to the pod
+		patchMetadata["annotations"] = patchAnnotations
+		if patchLabels != nil {
+			patchMetadata["labels"] = patchLabels
+		}
+		patch, _ := json.Marshal(map[string]interface{}{
+			"metadata": patchMetadata,
+		})
+		if err := r.Patch(context.TODO(), pod, client.RawPatch(types.StrategicMergePatchType, patch)); err != nil {
+			// no need to retry, this operation will be retried in the next reconcile.
+			// this failure would not affect the subsequent logic
+			klog.Errorf(`Failed to patch "matched-workloadspread:{Name: %s, Subset: %s} annotation" for pod (%s/%s), but will retry in the next reconcile, err: %s`,
```

Comment:
I'm not sure. If we failed to patch the pod, should we return the favoriteSubset with no error? (which means the pod has no annotation that belongs to the subset, but controller consider it does)

1. How to ensure the workloadspread will be triggered reconcile after this time?
2. What if the pod be allocated to another subset in next reconcile?

Reply:
emmm.... you're right,  returning this error can indeed get rid out of some problems.